# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/basetypes/HyperTextAccessibleBase.h
# Commit: a81a40b250b4
# Full Hash: a81a40b250b401d6133c2490d8d2d3bd52f1bc6d
# Author: James Teh <jteh@mozilla.com>
# Date: 2022-05-05 09:42:30
# Regressor Bug: 1766794
# File Overlap Count: 2
# Description:
#   Bug 1766794: Lazily cache HyperText offsets for RemoteAccessibles. r=eeejay
#   
#   We already had a similar cache in local HyperTextAccessible.
#   This improves performance significantly when walking the text attributes of a container with a large number of text leaf children, such as is encountered when using view source.
#   This patch unifies that cache across local and remote.
# ==============================================================================

diff -r 000ea1907a05 -r a81a40b250b4 accessible/basetypes/HyperTextAccessibleBase.h
--- a/accessible/basetypes/HyperTextAccessibleBase.h	Thu May 05 04:00:58 2022 +0000
+++ b/accessible/basetypes/HyperTextAccessibleBase.h	Thu May 05 04:31:12 2022 +0000
@@ -41,6 +41,12 @@
 class HyperTextAccessibleBase {
  public:
   /**
+   * Invalidate cached HyperText offsets. This should be called whenever a
+   * child is added or removed or the text of a text leaf child is changed.
+   */
+  virtual void InvalidateCachedHyperTextOffsets() = 0;
+
+  /**
    * Return child accessible at the given text offset.
    *
    * @param  aOffset  [in] the given text offset
@@ -59,17 +65,13 @@
    * accessible.
    *
    * @param  aChild           [in] accessible child to get text offset for
-   * @param  aInvalidateAfter [in, optional] indicates whether invalidate
-   *                           cached offsets for next siblings of the child
    */
-  int32_t GetChildOffset(const Accessible* aChild,
-                         bool aInvalidateAfter = false) const;
+  int32_t GetChildOffset(const Accessible* aChild) const;
 
   /**
    * Return text offset for the child accessible index.
    */
-  virtual int32_t GetChildOffset(uint32_t aChildIndex,
-                                 bool aInvalidateAfter = false) const;
+  virtual int32_t GetChildOffset(uint32_t aChildIndex) const;
 
   /**
    * Return character count within the hypertext accessible.
@@ -190,6 +192,20 @@
     return const_cast<Accessible*>(acc);
   }
 
+  /**
+   * Get the cached map of child indexes to HyperText offsets. If the cache
+   * hasn't been built yet, build it.
+   * This is an array which contains the exclusive end offset for each child.
+   * That is, the start offset for child c is array index c - 1.
+   */
+  virtual const nsTArray<int32_t>& GetCachedHyperTextOffsets() const = 0;
+
+  /**
+   * Build the HyperText offsets cache. This should only be called by
+   * GetCachedHyperTextOffsets.
+   */
+  void BuildCachedHyperTextOffsets(nsTArray<int32_t>& aOffsets) const;
+
  private:
   /**
    * Transform the given a11y point into an offset relative to this hypertext.